libs/isograph-react/src/core/reader.ts

```diff
@@ -17,32 +17,45 @@ export type TopLevelReaderArtifact<
   TReadFromStore extends Object,
   TClientFieldValue,
   TComponentProps extends Record<string, never>,
+  TVariables = Variables,
 > =
-  | EagerReaderArtifact<TReadFromStore, TClientFieldValue>
-  | ComponentReaderArtifact<TReadFromStore, TComponentProps>;
+  | EagerReaderArtifact<TReadFromStore, TClientFieldValue, TVariables>
+  | ComponentReaderArtifact<TReadFromStore, TComponentProps, TVariables>;
 
 export type EagerReaderArtifact<
   TReadFromStore extends Object,
   TClientFieldValue,
+  TVariables = Variables,
 > = {
   readonly kind: 'EagerReaderArtifact';
   readonly readerAst: ReaderAst<TReadFromStore>;
-  readonly resolver: (data: TReadFromStore) => TClientFieldValue;
+  readonly resolver: (
+    data: ResolverData<TReadFromStore, TVariables>,
+  ) => TClientFieldValue;
 };
 
 export type ComponentReaderArtifact<
   TReadFromStore extends Object,
   TComponentProps extends Record<string, unknown> = Record<string, never>,
+  TVariables = Variables,
 > = {
   readonly kind: 'ComponentReaderArtifact';
   readonly componentName: ComponentOrFieldName;
   readonly readerAst: ReaderAst<TReadFromStore>;
   readonly resolver: (
-    data: TReadFromStore,
+    data: ResolverData<TReadFromStore, TVariables>,
     runtimeProps: TComponentProps,
   ) => React.ReactNode;
 };
 
+export type ResolverData<
+  TReadFromStore extends Object,
+  TVariables = Variables,
+> = {
+  readonly data: TReadFromStore;
+  readonly variables: TVariables | null;
```

Comment:
I just pushed a commit that makes `variables` non-null in `FragmentReference`, so no need to make it null here